File: server/src/test/java/com/fluffy/reaction/domain/LikeQueryServiceTest.java
package com.fluffy.reaction.domain;

import static org.assertj.core.api.Assertions.assertThat;

import com.fluffy.auth.domain.Member;
import com.fluffy.auth.domain.MemberRepository;
import com.fluffy.support.AbstractIntegrationTest;
import com.fluffy.support.data.MemberTestData;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;

class LikeQueryServiceTest extends AbstractIntegrationTest {

    @Autowired
    private MemberRepository memberRepository;

    @Autowired
    private LikeQueryService likeQueryService;

    @Autowired
    private ReactionRepository reactionRepository;

    @Test
    @DisplayName("사용자가 특정 대상에 대해 좋아요를 했는지 여부를 조회할 수 있다.")
    void isLiked() {
        // given
        LikeTarget likeTarget = LikeTarget.EXAM;
        Long targetId = 1L;

        Member member = createMember();

        Reaction reaction = new Reaction(likeTarget.name(), targetId, member.getId(), ReactionType.LIKE);
        reactionRepository.save(reaction);

        // when
        boolean isLiked = likeQueryService.isLiked(new Like(likeTarget, targetId), member.getId());

        // then
        assertThat(isLiked).isTrue();
    }

    @Test
    @DisplayName("사용자가 특정 대상에 대해 좋아요를 하지 않았다면 거짓을 반환한다.")
    void isNotLiked() {
        // given
        LikeTarget likeTarget = LikeTarget.EXAM;
        Long targetId = 1L;

        Member member = createMember();

        // when
        boolean isLiked = likeQueryService.isLiked(new Like(likeTarget, targetId), member.getId());

        // then
        assertThat(isLiked).isFalse();
    }

    private Member createMember() {
        Member member = MemberTestData.defaultMember().build();
        return memberRepository.save(member);
    }
}

File: server/src/test/kotlin/com/fluffy/reaction/domain/LikeQueryServiceTest.kt
package com.fluffy.reaction.domain

import com.fluffy.global.web.Accessor
import io.kotest.core.spec.style.BehaviorSpec
import io.kotest.matchers.shouldBe
import io.mockk.every
import io.mockk.mockk

class LikeQueryServiceTest : BehaviorSpec({

    val reactionRepository = mockk<ReactionRepository>()
    val likeQueryService = LikeQueryService(reactionRepository)

    Given("isLiked") {
        val like = Like(LikeTarget.EXAM, 1L)
        val memberId = 1L

        When("사용자가 비회원인 경우") {
            val accessor = Accessor.GUEST
            val actual = likeQueryService.isLiked(like, accessor)

            Then("거짓을 반환한다.") {
                actual shouldBe false
            }
        }

        When("사용자가 회원인 경우") {
            val accessor = Accessor(memberId)

            And("좋아요가 활성 상태인 경우") {
                val reaction = Reaction(
                    targetType = like.target.name,
                    targetId = like.targetId,
                    memberId = memberId,
                    type = ReactionType.LIKE,
                    status = ReactionStatus.ACTIVE
                )

                every {
                    reactionRepository.findByTargetTypeAndTargetIdAndMemberIdAndType(any(), any(), any(), any())
                } returns reaction

                val actual = likeQueryService.isLiked(like, accessor)

                Then("참을 반환한다.") {
                    actual shouldBe true
                }
            }

            And("좋아요가 삭제 상태인 경우") {
                val reaction = Reaction(
                    targetType = like.target.name,
                    targetId = like.targetId,
                    memberId = memberId,
                    type = ReactionType.LIKE,
                    status = ReactionStatus.DELETED
                )

                every {
                    reactionRepository.findByTargetTypeAndTargetIdAndMemberIdAndType(any(), any(), any(), any())
                } returns reaction

                val actual = likeQueryService.isLiked(like, accessor)

                Then("거짓을 반환한다.") {
                    actual shouldBe false
                }
            }

            And("좋아요가 존재하지 않는 경우") {
                every {
                    reactionRepository.findByTargetTypeAndTargetIdAndMemberIdAndType(any(), any(), any(), any())
                } returns null

                val actual = likeQueryService.isLiked(like, accessor)

                Then("거짓을 반환한다.") {
                    actual shouldBe false
                }
            }
        }
    }
})
